markdown.dtx

```diff
@@ -14915,6 +14923,12 @@ function M.writer.new(options)
     local name = util.cache(options.cacheDir, s, nil, nil, ".verbatim")
     return {"\\markdownRendererInputFencedCode{",name,"}{",i,"}"}
   end
+  function self.jekyllData(s)
+    if not self.is_writing then return "" end
+    s = string.gsub(s, '([^:]+):%s([^\r\n]+)[\r\n]+', '\\markdownRendererSetJekyllData{%1}{%2}')
+    s = string.gsub(s, '[\r\n%s]*$','')
```

Comment:
This seems quite rudimentary. We would definitely like to support at least continued lines, lists, and structures.

### References
1. https://en.wikipedia.org/wiki/YAML
2. https://yaml.org/spec/
3. https://pandoc.org/MANUAL.html#extension-yaml_metadata_block

Reply:
What this ultimately means is that we'll want to parse YAML using LPEG instead of reading the entire block into a string and then parsing it with regexes.

Reply:
However, this warrants some consideration of the API, since we want to avoid future changes to the API whenever possible.
Therefore, we should decide how the following document should parse:
``` yaml
title: The document title
author:
- name: Author One
  affiliation: University of Somewhere
- name: Author Two
  affiliation: University of Nowhere
```

Here is one idea, where `...` can be `JekyllData`, `FrontMatter`, `YAMLMetadata`, or some other appropriate name:
``` tex
\markdownRenderer...Begin
\markdownRenderer...String{title}{The document title}
\markdownRenderer...SequenceBegin{author}{2}
\markdownRenderer...MappingBegin{1}{2}
\markdownRenderer...String{name}{Author One}
\markdownRenderer...String{affiliation}{University of Somewhere}
\markdownRenderer...MappingEnd
\markdownRenderer...MappingBegin{2}{2}
\markdownRenderer...String{name}{Author Two}
\markdownRenderer...String{affiliation}{University of Nowhere}
\markdownRenderer...MappingEnd
\markdownRenderer...SequenceEnd
\markdownRenderer...End
```

The individual renderers would have the following semantics:

- `\markdownRenderer...Begin`: The beginning of a YAML metadata block
- `\markdownRenderer...End`: The end of a YAML metadata block
- `\markdownRenderer...MappingBegin{`*key*`}{`*length*`}`: The beginning of a mapping at the key/index *key* at the current nesting level, with *length* elements in the mapping.
- `\markdownRenderer...MappingEnd`: The end of a mapping.
- `\markdownRenderer...SequenceBegin{`*key*`}{`*length*`}`: The beginning of a sequence at the key/index *key* at the current nesting level, with *length* elements in the sequence.
- `\markdownRenderer...SequenceEnd`: The end of a sequence.
- `\markdownRenderer...String{`*key*`}{`*value*`}`: A string at the key/index *key* with the value *value* at the current nesting level.

Reply:
So i mostly did this, just have to take some closer look at the order of the definitions inside the package. just placed it at the same spots as before. If you are faster with this, feel free, because I am not as used to your structure ;-) 

Additionally I am still thinking about what would be a good start for a default behaviour.

Reply:
In LaTeX, we could create a `keyval` namespace:

``` latex
\markdownSetup{
  rendererPrototypes = {
    jekyllDataString = {\setkeys{markdownLaTeXJekyllData}{#1=#2}},
  }
}
```

Then, we could set some sane defaults such as:

``` latex
\define@key{markdownLaTeXJekyllData}{author}{\author{#1}}
\define@key{markdownLaTeXJekyllData}{title}{\title{#1}}
\define@key{markdownLaTeXJekyllData}{date}{\date{#1}}
```

In order to ignore unknown keys by default, we could have a setup similar to this:

``` latex
\markdownSetup{
  rendererPrototypes = {
    jekyllDataBegin = {%
      \begingroup
        \let\KV@errx\@gobble
    }, jekyllDataEnd = {%
        \maketitle
      \endgroup
    },
  }
}
```

Reply:
I'd prefer l3keys, since you want to modernize the syntax as mentioned in #96 I think it would be simpler to use a the l3prefix.

Reply:
@TeXhackse There are several prefixes registered: `markdownOptions`, `markdownRenderers` and `markdownRendererPrototypes`. However, these are all an implementation detail and shall be replaced with `l3keys` further down the road.

Reply:
@Witiko Yep I got that. But there is currently no expl3 prefix registered for markdown, is it? 

I wonder if it's not a good idea to register that one and start this part directly in expl3. The other parts can be moved from keyval later on. But I find it a but inefficient to start this one in keyval if a move towards expl3 is planned anyway … 

There on would use `markdown` and for this I'd then use `markdown/jekyllData` or `markdown/data`

Reply:
I agree, `markdown/jekyllData` sounds as the obvious first choice.

Reply:
However, our first priority is making the unit tests pass. Having the default renderer prototype for LaTeX use `l3keys` is just a cherry on top.